File: seqnone/sequence.py
import liblo

class Event(liblo.Message):

    def __init__(self, address, value):

        liblo.Message.__init__(self, address, value)
        self.value = value

class Sequence():

    def __init__(self, parent, name, length, address, events, enabled=False, note=False):

        self.parent = parent
        self.name = name
        self.length = length
        self.address = address

        self.events = {}
        for i in events:
            self.events[i] = Event(self.address, events[i])

        self.note = bool(note)
        self.note_active = False
        if self.note:
            self.note_off = Event(self.address, 0)

        self.enabled = bool(enabled)
        self.toggle(self.enabled)



    def play(self, cursor):
        
        i = cursor % self.length

        if i in self.events:

            self.parent.send(self.events[i])

            if self.note:

                self.note_active = bool(self.events[i].value)

    def enable(self):

        self.toggle(True)

    def disable(self):

        self.toggle(False)

    def toggle(self, state):

        if self.note_active and self.enabled and not state:

            self.parent.send(self.note_off)


        if state and self not in self.parent.active_sequences:
            self.parent.active_sequences.append(self)

        if not state and self in self.parent.active_sequences:
            del self.parent.active_sequences[self.parent.active_sequences.index(self)]

        self.enabled = state

File: src/sequence.cpp
#include <algorithm>

#include "sequence.hpp"
#include "sequencer.hpp"

Sequence::Sequence(){} // empty default constructor for std::map

Sequence::Sequence(Sequencer *seq, std::string osc_address, const char* osc_type, ValueMap v, int seq_length, bool state, bool is_note)
{

    sequencer = seq;

    address = osc_address;

    type = osc_type;

    if (type[0] !=  'i' && type[0] !=  'f' && type[0] !=  'd') {
        type = "i";
    }

    values = v;
    length = seq_length;

    note = is_note;
    note_on = false;

    if (state) enable();

    feed_status(false);


}

Sequence::~Sequence()
{

    disable();
    feed_status(true);

}

void Sequence::enable()
{

    SequenceVectorIterator it = find(sequencer->sequence_active.begin(), sequencer->sequence_active.end(), this);

    if (it == sequencer->sequence_active.end()) {
        sequencer->sequence_active.push_back(this);
    }

    enabled = true;
    feed_status(false);

}

void Sequence::disable()
{

    note_off();

    SequenceVectorIterator it = find(sequencer->sequence_active.begin(), sequencer->sequence_active.end(), this);

    if (it != sequencer->sequence_active.end()) {
        sequencer->sequence_active.erase(it);
    }

    enabled = false;
    feed_status(false);

}

void Sequence::toggle()
{

    if (enabled) {
        disable();
    } else {
        enable();
    }

}

void Sequence::play(int c)
{

    if (!enabled || !length) return;

    int cursor = c % length;

    if (values.find(cursor) != values.end()) {

        sequencer->osc_send(address, type, values[cursor]);

        if (note) {
            note_on = values[cursor] > 0;
        }

    }

}

void Sequence::note_off()
{

    if (!enabled || !note_on) return;

    note_on = false;
    sequencer->osc_send(address, type, 0);

}

void Sequence::feed_status(bool deleted)
{

    std::string json = "{";

    json += "\"address\":\"" + address + "\",";

    if (deleted) {

        json += "\"removed\":true";

    } else {

        json += "\"enabled\":" + std::to_string(enabled) + ",";
        json += "\"type\":\"" + (std::string)type + "\",";
        json += "\"note\":" + std::to_string(note) + ",";
        json += "\"length\":" + std::to_string(length) + ",";
        // if (note) json += "\"note_on\":" + std::to_string(note_on) + ",";

        json += "\"values\":{" ;

        for (auto it2 = values.cbegin(); it2 != values.cend();) {
            json += "\"" + std::to_string(it2->first) + "\":" + std::to_string(it2->second);
            it2++;
            if (it2 != values.cend()) json += ",";
        }

        json += "}";

    }

    json += "}";

    sequencer->osc_send_feed("/status/sequence", json);

}
